File: cpp/198.house-robber.cpp
/*
 * @lc app=leetcode id=198 lang=cpp
 *
 * [198] House Robber
 */
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    int rob(vector<int> &nums)
    {
        int len = nums.size();
        if (len < 3)
        {
            int res = nums[0];
            for (int i = 1; i != len; ++i)
            {
                res = max(res, nums[i]);
            }
            return res;
        }
        deque<int> ts = {nums[0], nums[1], nums[0] + nums[2]};
        for (int i = 3; i != len; ++i)
        {
            ts.push_back(nums[i] + max(ts[0], ts[1]));
            ts.pop_front();
        }
        return max(ts[2], ts[1]);
    }
};
// @lc code=end

File: 198.house-robber.py
#
# @lc app=leetcode id=198 lang=python3
#
# [198] House Robber
#
from typing import *
# @lc code=start


class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = [nums[0], 0]
        for i in range(1, len(nums)):
            dp[0], dp[1] = dp[1]+nums[i], max(dp)
        return max(dp)

# @lc code=end
